x/params/keeper.go

```diff
@@ -1,405 +1,112 @@
-	bz := store.Get([]byte(key))
-	err = k.k.cdc.UnmarshalBinary(bz, &res)
-	return
-}
-
-// GetUint64 is helper function for uint64 params
-func (k Getter) GetUint64(ctx sdk.Context, key string) (res uint64, err error) {
-	store := ctx.KVStore(k.k.key)
-	bz := store.Get([]byte(key))
-	err = k.k.cdc.UnmarshalBinary(bz, &res)
-	return
-}
-
-// GetInt is helper function for sdk.Int params
-func (k Getter) GetInt(ctx sdk.Context, key string) (res sdk.Int, err error) {
-	store := ctx.KVStore(k.k.key)
-	bz := store.Get([]byte(key))
-	err = k.k.cdc.UnmarshalBinary(bz, &res)
-	return
-}
-
-// GetUint is helper function for sdk.Uint params
-func (k Getter) GetUint(ctx sdk.Context, key string) (res sdk.Uint, err error) {
-	store := ctx.KVStore(k.k.key)
-	bz := store.Get([]byte(key))
-	err = k.k.cdc.UnmarshalBinary(bz, &res)
-	return
-}
-
-// GetRat is helper function for rat params
-func (k Getter) GetRat(ctx sdk.Context, key string) (res sdk.Rat, err error) {
-	store := ctx.KVStore(k.k.key)
-	bz := store.Get([]byte(key))
-	err = k.k.cdc.UnmarshalBinary(bz, &res)
-	return
-}
-
-// GetStringWithDefault is helper function for string params with default value
-func (k Getter) GetStringWithDefault(ctx sdk.Context, key string, def string) (res string) {
-	store := ctx.KVStore(k.k.key)
-	bz := store.Get([]byte(key))
-	if bz == nil {
-		return def
-	}
-	k.k.cdc.MustUnmarshalBinary(bz, &res)
-	return
-}
-
-// GetBoolWithDefault is helper function for bool params with default value
-func (k Getter) GetBoolWithDefault(ctx sdk.Context, key string, def bool) (res bool) {
-	store := ctx.KVStore(k.k.key)
-	bz := store.Get([]byte(key))
-	if bz == nil {
-		return def
-	}
-	k.k.cdc.MustUnmarshalBinary(bz, &res)
-	return
-}
+	cdc  *wire.Codec
+	key  sdk.StoreKey
+	tkey sdk.StoreKey
 
-// GetInt16WithDefault is helper function for int16 params with default value
-func (k Getter) GetInt16WithDefault(ctx sdk.Context, key string, def int16) (res int16) {
-	store := ctx.KVStore(k.k.key)
-	bz := store.Get([]byte(key))
-	if bz == nil {
-		return def
-	}
-	k.k.cdc.MustUnmarshalBinary(bz, &res)
-	return
-}
+	space KeeperSpace
 
-// GetInt32WithDefault is helper function for int32 params with default value
-func (k Getter) GetInt32WithDefault(ctx sdk.Context, key string, def int32) (res int32) {
-	store := ctx.KVStore(k.k.key)
-	bz := store.Get([]byte(key))
-	if bz == nil {
-		return def
-	}
-	k.k.cdc.MustUnmarshalBinary(bz, &res)
-	return
+	stores map[string]*Store
 }
 
-// GetInt64WithDefault is helper function for int64 params with default value
-func (k Getter) GetInt64WithDefault(ctx sdk.Context, key string, def int64) (res int64) {
-	store := ctx.KVStore(k.k.key)
-	bz := store.Get([]byte(key))
-	if bz == nil {
-		return def
-	}
-	k.k.cdc.MustUnmarshalBinary(bz, &res)
-	return
+// KeeperSpace defines param space for the substores
+// Zero value("") means not using the substore
+type KeeperSpace struct {
+	ConsensusSpace string
+	GasConfigSpace string
+	MsgStatusSpace string
```

Comment:
* Lets turn the KeeperSpace keys as a "paramSpaceKey".

```
// NOTE: Doesn't need to be an interface as in types/store.go StoreKey,
// since most likely there will only be 1 implementation of a param subspace, unlike IAVLStore, PatriciaStore, TransientStore etc.

// ParamSpaceKey is a key used to index ....
type ParamSpaceKey struct {
    key string // alpha numeric
}
```

Reply:
Is paramSpaceKey necessary here? Space strings are just for distinguishing namespace between paramstore users, so I think subspace generation allowing only unique spacename provides enough security. If we introduce the storeKey here, it could increase complexity, considering that the main gov module has to access all the subspaces.